docs/best_practices/python.md

```diff
@@ -0,0 +1,167 @@
+# Python best practices
+
+## Import
+
+### Avoid `from module_foo import symbol_bar`
+
+In general, it is a good practice to avoid the form `from foo import bar`
+because it introduces two distinct bindings (`bar` is distinct from `foo.bar`)
+and when the binding in one namespace changes, the binding in the other will
+not…
+
+That's also why this can interfere with the mocking.
+
+All in all, this should be avoided when unecessary.
+
+#### Rationale
+
+Reduce the likelihood of surprising behaviors and ease the mocking.
+
+#### Example
+
+``` python
+# Good
+import foo
+
+baz = foo.Bar()
+
+# Bad
+from foo import Bar
+
+baz = Bar()
+```
+
+#### References
+
+- [Idioms and Anti-Idioms in Python](https://docs.python.org/3.1/howto/doanddont.html#from-module-import-name1-name2%0A)
+- [unittest.mock documentation](https://docs.python.org/3.6/library/unittest.mock.html#where-to-patch)
+
+## Naming
+
+### Predicate functions
+
+Functions that return a Boolean value should have a name that starts with
+`has_`, `is_`, `was_`, `can_` or something similar that makes it clear that it's
+a Boolean (or returns one).
+
+This recommandation also applies to Boolean variable.
+
+#### Rationale
+
+Makes code clearer and more expressive.
+
+#### Example
+
+``` python
+class Foo:
+    # Bad.
+    def empty(self):
+        return len(self.bar) == 0
+
+    # Bad.
+    def baz(self, initialized):
+        if initialized:
+            return
+        […]
+
+    # Good.
+    def is_empty(self):
+        return len(self.bar) == 0
+
+    # Good.
+    def qux(self, is_initialized):
+        if is_initialized:
+            return
+        […]
+```
+
+## Patterns and idioms
+
+### Don't write code vulnerable to "Time of check to time of use"
+
+When there is a time windows between the checking of a condition and the use of
+the result of that check where the result may become outdated, you should always
+follow the the **EAFP** (It is easier to ask for forgiveness than permission)
+philosphopy rather than the **LBYL** (look before you leap) one (because it
+gives you a false sense of security).
+
+Otherwise, your code will be vulnerable to the infamous **TOCTTOU** (Time of
+check to time of use) bugs.
+
+In Python terms:
+- **LBYL**: `if` guard around the action
+- **EAFP**: `try`/`except` around the action
```

Comment:
```suggestion
- **EAFP**: `try`/`except` statement around the action
```

Reply:
Done (but I used `statements` as I think `try` and `except` counts as two statements)